src/pyodide/internal/workers.py

```diff
@@ -149,54 +157,211 @@ def json(
+        for val in self._js_form_data.values():
+            yield _js_value_to_py(val)
 
     def keys(self) -> Generator[str, None, None]:
-        yield from self.js_form_data.keys()
+        yield from self._js_form_data.keys()
 
     def __iter__(self):
         yield from self.keys()
 
     def items(self) -> Generator[tuple[str, FormDataValue], None, None]:
-        for item in self.js_form_data.entries():
-            yield (item[0], item[1])
+        for k, v in self._js_form_data.entries():
+            yield (k, _js_value_to_py(v))
 
     def __delitem__(self, key: str):
         self.delete(key)
 
     def __len__(self):
         return len(self.keys())
+
+    def get_all(self, key: str) -> list[FormDataValue]:
+        return list(map(lambda x: _js_value_to_py(x), self._js_form_data.getAll(key)))
+
+    @property
+    def js_object(self) -> "js.FormData":
+        return self._js_form_data
+
+
+def _supports_buffer_protocol(o):
+    try:
+        # memoryview used only for testing type; 'with' releases the view instantly
+        with memoryview(o):
+            return True
+    except TypeError:
+        return False
+
+
+@contextmanager
+def _make_blob_entry(e):
+    if isinstance(e, str):
+        yield e
+        return
+    if isinstance(e, Blob):
+        yield e._js_blob
+        return
+    if hasattr(e, "constructor") and (e.constructor.name in ("Blob", "File")):
+        yield e
+        return
+    if _supports_buffer_protocol(e):
+        px = create_proxy(e)
+        buf = px.getBuffer()
+        try:
+            yield buf.data
+            return
+        finally:
+            buf.release()
+            px.destroy()
+    raise TypeError(f"Don't know how to handle {type(e)} for Blob()")
+
+
+def _is_iterable(obj):
+    try:
+        iter(obj)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+BlobValue = (
+    "str | bytes | js.ArrayBuffer | js.TypedArray | js.DataView | js.Blob | Blob"
+)
+
+
+class BlobEnding(StrEnum):
+    TRANSPARENT = "transparent"
+    NATIVE = "native"
+
+
+class Blob:
+    def __init__(
+        self,
+        blob_parts: "Iterable[BlobValue] | BlobValue",
+        content_type: str | None = None,
+        endings: BlobEnding | str | None = None,
+    ):
+        if endings:
+            endings = str(endings)
+
+        is_single_item = not _is_iterable(blob_parts)
+        if is_single_item:
+            if not content_type:
+                # Inherit the `type` if we have a single item.
+                self._js_blob = (
+                    blob_parts.js_object if isinstance(blob_parts, Blob) else blob_parts
+                )
+                return
+
+            # Otherwise create a new Blob below.
+            blob_parts = [blob_parts]
+
+        with ExitStack() as stack:
+            args = [stack.enter_context(_make_blob_entry(e)) for e in blob_parts]
+            self._js_blob = js.Blob.new(to_js(args), type=content_type, endings=endings)
```

Comment:
We need to be careful here or else this will be leaky. Probably the easiest thing to do is:
```suggestion
            pyproxies = js.Array()
            self._js_blob = js.Blob.new(to_js(args, pyproxies), type=content_type, endings=endings)
            destroy_proxies(pyproxies)
```
you need also a `from pyodide.ffi import to_js`.

Reply:
Yeah we definitely don't need this here sorry. Everything returned from `_make_blob_entry` is either a string or a JsProxy so no PyProxies are created by this to_js call.